### PR TITLE
RO export in HTML and Etherpad format should not expose the writeable pad id

### LINKED ISSUE #5094: 🚨 Potential Security Vulnerability
Hello, @JohnMcLear @Pita   @rhansen @marcelklehr  - a potential high severity security vulnerability in your repository has been disclosed to huntr.

Visit report url and validate them 
https://huntr.dev/bounties/1624811768438-ether/etherpad-lite/

Report details is only visible to repo-maintainer and reporter

### PR DESCRIPTION
**Pull request by webzwo0i**
tests not included yet

Fixes #5094

### CODE REVIEW

File: src/node/hooks/express/padreadonly.js [hunk 1]
```diff
@@ -19,7 +19,7 @@ exports.expressCreateServer = (hookName, args, cb) => {
 
     if (await hasPadAccess(req, res)) {
       // render the html document
-      const html = await exporthtml.getPadHTMLDocument(padId, null);
+      const html = await exporthtml.getPadHTMLDocument(padId, null, null);
```
Comment (rhansen) [inline src/node/hooks/express/padreadonly.js:22]: Shouldn't this be:
```suggestion
      const html = await exporthtml.getPadHTMLDocument(padId, null, req.params.id);
```

File: src/node/utils/ExportEtherpad.js [hunk 2]
```diff
@@ -63,7 +69,11 @@ exports.getPadRaw = async (padId) => {
   const prefixes = await hooks.aCallAll('exportEtherpadAdditionalContent');
   await Promise.all(prefixes.map(async (prefix) => {
     const key = `${prefix}:${padId}`;
-    data[key] = await db.get(key);
+    if (readOnlyId) {
+      data[`${prefix}:${readOnlyId}`] = await db.get(key);
+    } else {
+      data[key] = await db.get(key);
+    }
```
Comment (rhansen) [inline src/node/utils/ExportEtherpad.js:76]:
```suggestion
    const writeKey = readOnlyId ? `${prefix}:${readOnlyId}` : key;
    data[writeKey] = await db.get(key);
```

File: src/node/utils/ExportEtherpad.js [hunk 3]
```diff
@@ -50,7 +56,7 @@ exports.getPadRaw = async (padId) => {
           if (authorEntry) {
             data[`globalAuthor:${authorId}`] = authorEntry;
             if (authorEntry.padIDs) {
-              authorEntry.padIDs = padId;
+              authorEntry.padIDs = readOnlyId ? readOnlyId : padId;
```
Comment (rhansen) [inline src/node/utils/ExportEtherpad.js:59]:
```suggestion
              authorEntry.padIDs = readOnlyId || padId;
```

File: src/node/utils/ExportHtml.js [hunk 4]
```diff
@@ -475,7 +475,7 @@ exports.getPadHTMLDocument = async (padId, revNum) => {
 
   return eejs.require('ep_etherpad-lite/templates/export_html.html', {
     body: html,
-    padId: Security.escapeHTML(padId),
+    padId: readOnlyId ? Security.escapeHTML(readOnlyId) : Security.escapeHTML(padId),
```
Comment (rhansen) [inline src/node/utils/ExportHtml.js:478]:
```suggestion
    padId: Security.escapeHTML(readOnlyId || padId),
```

File: src/node/hooks/express/padreadonly.js (hunk 1 above)
Comment (webzwo0i) [reply src/node/hooks/express/padreadonly.js:22]: Good catch! It's leaking the writeable id without your change. However, it turns out the whole file seems to be dead code and non-working anyway - `null` as revision parameter is true in this [branch](https://github.com/ether/etherpad-lite/blob/develop/src/node/utils/ExportHtml.js#L32) and `null/100` equals 0 in https://github.com/ether/etherpad-lite/blob/develop/src/node/db/Pad.js#L237 so it's always returning the 0 revision.

`/ro/:id` is no longer used, as we access both writeable and readonly ids via `/p/:id` now. So I'm going to remove the whole thing

File: src/node/utils/ExportEtherpad.js [hunk 5]
```diff
@@ -35,7 +35,13 @@ exports.getPadRaw = async (padId) => {
...
   for (const key of records) {
     // For each piece of info about a pad.
-    const entry = data[key] = await db.get(key);
+    const entry = await db.get(key);
+
+    if (readOnlyId) {
+      data[key.replace(padKey, `pad:${readOnlyId}`)] = entry;
+    } else {
+      data[key] = entry;
+    }
```
Comment (rhansen) [inline src/node/utils/ExportEtherpad.js:44]:
```suggestion
    const writeKey = readOnlyId ? key.replace(padKey, `pad:${readOnlyId}`) : key;
    const entry = data[writeKey] = await db.get(key);
```

**Comment by rhansen** [comment]
What's the state of this @webzwo0i?

**Comment by JohnMcLear** [comment]
bump @webzwo0i

**Comment by JohnMcLear** [comment]
tests now included.

**Comment by rhansen** [comment]
I rebased and made some changes:
  * Reworded commit messages.
  * Squashed the two fix commits into one.
  * Nuked the empty "trigger tests" commit.
  * Fixed and restructured the tests.
  * Added more tests.

**Pull request merged**